File: desktop/src/main/telemetry-health.ts
import { Notification } from 'electron'
import { log, warn } from './logger'
import type { EnterprisePolicy } from '../shared/types-enterprise'

/**
 * Desktop consumer for the engine's `engine_telemetry_health` event
 * (issue #379).
 *
 * The engine reports the delivery health of each telemetry egress target and
 * has no opinion about what a consumer does with it. The desktop's opinion:
 * an operator whose audit stream has stopped arriving should be told while
 * the backlog is still draining, not when a downstream query later returns
 * nothing.
 *
 * Notifications are deliberately sparse. The engine already suppresses
 * steady-state growth and only reports threshold crossings, recovery, and
 * critical conditions, so this module notifies on what it receives rather
 * than adding a second layer of rate limiting — with one exception noted at
 * `shouldNotify`.
 */

/** Shape of the fields this consumer reads off the engine event. */
export interface TelemetryHealthEvent {
  type?: string
  telemetryTarget?: string
  telemetryQueuedEvents?: number
  telemetryQueuedBytes?: number
  telemetryOldestAgeMs?: number
  telemetrySoftWarnBytes?: number
  telemetryPercentOfSoftWarn?: number
  telemetryCrossedThreshold?: number
  telemetryHealthy?: boolean
  telemetryLastError?: string
  telemetryCritical?: boolean
  telemetryStuck?: boolean
  telemetryStuckAfterMs?: number
  telemetryMaxAttempts?: number
  telemetryQuarantinedEvents?: number
  telemetryQuarantinedBytes?: number
}

/** Latest known state per target, for a surface that wants to render it. */
export interface TelemetryHealthState {
  target: string
  queuedEvents: number
  queuedBytes: number
  oldestAgeMs: number
  percentOfSoftWarn: number
  healthy: boolean
  lastError?: string
  critical: boolean
  stuck: boolean
  maxAttempts: number
  quarantinedEvents: number
  quarantinedBytes: number
  updatedAt: number
}

const state = new Map<string, TelemetryHealthState>()

/** Current health of every target the engine has reported on. */
export function telemetryHealthState(): TelemetryHealthState[] {
  return [...state.values()]
}

/**
 * Whether this observation warrants interrupting the operator.
 *
 * A crossing, a critical condition, a newly stuck queue, or newly
 * quarantined events do. A recovery does only if we had previously warned
 * about that target — telling someone their telemetry recovered when they
 * were never told it was failing is noise, not news.
 */
function shouldNotify(evt: TelemetryHealthEvent, prev: TelemetryHealthState | undefined, hadWarned: boolean): boolean {
  if (evt.telemetryCritical) return true
  if ((evt.telemetryCrossedThreshold ?? 0) > 0) return true
  if (evt.telemetryStuck && !prev?.stuck) return true
  if ((evt.telemetryQuarantinedEvents ?? 0) > (prev?.quarantinedEvents ?? 0)) return true
  if (evt.telemetryHealthy && !evt.telemetryStuck && hadWarned) return true
  return false
}

function formatAge(ms: number): string {
  const minutes = Math.round(ms / 60_000)
  if (minutes >= 120) return `${Math.round(minutes / 60)} hours`
  if (minutes >= 1) return `${minutes} minute${minutes === 1 ? '' : 's'}`
  return `${Math.round(ms / 1000)} seconds`
}

function formatBytes(bytes: number): string {
  if (bytes >= 1024 * 1024) return `${Math.round(bytes / (1024 * 1024))} MB`
  if (bytes >= 1024) return `${Math.round(bytes / 1024)} KB`
  return `${bytes} B`
}

/**
 * The operator-facing sentence for one observation. `prev` is the last state
 * seen for the target, used to tell a fresh quarantine apart from a stale
 * count riding along on an unrelated report.
 */
export function describeTelemetryHealth(evt: TelemetryHealthEvent, prev?: TelemetryHealthState): string {
  const target = evt.telemetryTarget || 'telemetry'
  if (evt.telemetryCritical) {
    return `Ion cannot save undelivered ${target} telemetry to disk (${evt.telemetryLastError ?? 'write failed'}). Events are being lost now.`
  }
  const quarantined = evt.telemetryQuarantinedEvents ?? 0
  if (quarantined > (prev?.quarantinedEvents ?? 0)) {
    const fresh = quarantined - (prev?.quarantinedEvents ?? 0)
    return `Ion could not send ${fresh} ${target} event${fresh === 1 ? '' : 's'} because ${fresh === 1 ? 'it is' : 'they are'} larger than the destination accepts. ${fresh === 1 ? 'It is' : 'They are'} kept on this machine in the quarantine file but ${fresh === 1 ? 'is' : 'are'} missing from the stream.`
  }
  if (evt.telemetryStuck) {
    const queued = evt.telemetryQueuedEvents ?? 0
    const attempts = evt.telemetryMaxAttempts ?? 0
    return `Ion's ${target} telemetry has been stuck for ${formatAge(evt.telemetryOldestAgeMs ?? 0)}. ${queued} event${queued === 1 ? '' : 's'} queued, ${attempts} delivery attempt${attempts === 1 ? '' : 's'} so far${evt.telemetryLastError ? ` (${evt.telemetryLastError})` : ''}. Events are preserved and will keep being retried.`
  }
  if (evt.telemetryHealthy) {
    return `Ion's ${target} telemetry backlog has cleared. Queued events were delivered.`
  }
  const queued = evt.telemetryQueuedEvents ?? 0
  const size = formatBytes(evt.telemetryQueuedBytes ?? 0)
  const pct = evt.telemetryPercentOfSoftWarn ?? 0
  return `Ion's ${target} telemetry is not being delivered. ${queued} event${queued === 1 ? '' : 's'} queued (${size}, ${pct}% of the warning threshold). Events are preserved and will be resent.`
}

/**
 * Whether telemetry-health notifications are permitted under the given
 * enterprise policy. Absent policy, or an absent/false flag, permits them —
 * the safe default for every unmanaged install. Exported as a pure function
 * (rather than inlined at the `installTelemetryHealthConsumer` call site) so
 * the enterprise-seal behavior is testable without standing up the real
 * engine bridge or enterprise-policy cache.
 */
export function telemetryHealthNotificationsEnabled(policy: EnterprisePolicy | null | undefined): boolean {
  return !policy?.disableTelemetryHealthNotifications
}

/**
 * Wire the consumer to the engine bridge.
 *
 * `enabled` is read per-event rather than captured, so an enterprise policy
 * that turns notifications off takes effect immediately instead of at the
 * next restart.
 */
export function installTelemetryHealthConsumer(
  engineBridge: { on: (ev: string, cb: (key: string, event: TelemetryHealthEvent) => void) => void },
  opts: { enabled: () => boolean } = { enabled: () => true },
): void {
  const warned = new Set<string>()

  engineBridge.on('event', (_key: string, event: TelemetryHealthEvent) => {
    if (event.type !== 'engine_telemetry_health') return

    const target = event.telemetryTarget || 'unknown'
    const prev = state.get(target)
    state.set(target, {
      target,
      queuedEvents: event.telemetryQueuedEvents ?? 0,
      queuedBytes: event.telemetryQueuedBytes ?? 0,
      oldestAgeMs: event.telemetryOldestAgeMs ?? 0,
      percentOfSoftWarn: event.telemetryPercentOfSoftWarn ?? 0,
      healthy: event.telemetryHealthy ?? false,
      lastError: event.telemetryLastError,
      critical: event.telemetryCritical ?? false,
      stuck: event.telemetryStuck ?? false,
      maxAttempts: event.telemetryMaxAttempts ?? 0,
      quarantinedEvents: event.telemetryQuarantinedEvents ?? 0,
      quarantinedBytes: event.telemetryQuarantinedBytes ?? 0,
      updatedAt: Date.now(),
    })

    const message = describeTelemetryHealth(event, prev)
    // Always logged, notification or not: an operator with notifications
    // disabled still needs the record, and this is the only desktop-side
    // trace that the engine reported a delivery problem at all.
    const fields = {
      target,
      healthy: event.telemetryHealthy ?? false,
      critical: event.telemetryCritical ?? false,
      queued_events: event.telemetryQueuedEvents ?? 0,
      queued_bytes: event.telemetryQueuedBytes ?? 0,
      percent_of_soft_warn: event.telemetryPercentOfSoftWarn ?? 0,
      crossed_threshold: event.telemetryCrossedThreshold ?? 0,
      stuck: event.telemetryStuck ?? false,
      oldest_age_ms: event.telemetryOldestAgeMs ?? 0,
      max_attempts: event.telemetryMaxAttempts ?? 0,
      quarantined_events: event.telemetryQuarantinedEvents ?? 0,
    }
    if (event.telemetryCritical || event.telemetryStuck || !event.telemetryHealthy) {
      warn('telemetry_health', 'telemetry delivery degraded', fields)
    } else {
      log('telemetry_health', 'telemetry delivery healthy', fields)
    }

    const hadWarned = warned.has(target)
    if (!shouldNotify(event, prev, hadWarned)) return
    if (event.telemetryHealthy && !event.telemetryStuck) {
      warned.delete(target)
    } else {
      warned.add(target)
    }

    if (!opts.enabled()) return
    if (!Notification.isSupported()) return
    new Notification({ title: 'Ion', body: message }).show()
  })
}

/** Test hook: drop retained state between cases. */
export function __resetTelemetryHealthForTest(): void {
  state.clear()
}

File: engine/internal/telemetry/telemetry_health.go
package telemetry

import (
	"errors"
	"io/fs"
	"sync"
	"syscall"
	"time"
)

// telemetry_health.go surfaces the delivery health of a target's retry queue
// as observable state (issue #379, and the "telemetry on whether telemetry is
// flowing" requirement in Orion 0008).
//
// The engine's own logs already record every delivery failure, but a log line
// is a poor foundation for an operator alert: it requires someone to be
// tailing, and it carries no notion of "how bad is this now" versus "how bad
// was it a minute ago". A queue that is quietly growing during a multi-hour
// outage looks identical, line by line, to one that failed twice and
// recovered.
//
// So the queue reports structured state, and a host wires an observer to turn
// that into whatever its consumers need — a typed engine event, a desktop
// notification, an OTEL gauge. The telemetry package deliberately does not
// know which: it is a leaf package with no business importing the event bus,
// the same reason ConversationEmitter takes a BeforeEventFunc rather than
// calling hooks itself.
//
// Three things escalate, independently: backlog size (against the soft-warn
// threshold), backlog age (a queue whose oldest batch has waited past the
// stuck threshold, at any size), and quarantine (content that left the
// stream). The second exists because the first once stayed silent through
// 255 failed attempts on a batch that was small but undeliverable.

// warnFractions are the fractions of the soft-warning threshold that escalate
// the health signal. An operator watching a backlog wants to hear about it
// while it is still draining, not at the moment the disk fills, so the first
// crossing is deliberately early (half the advisory threshold).
var warnFractions = []int{50, 75, 85, 95}

// TelemetryHealth is one target's delivery state at a point in time.
//
// It is a snapshot, not a delta: every field describes the queue as it stands
// when the observation is made, so a consumer replaces its view rather than
// accumulating. This matches the engine's existing snapshot-event convention
// (see docs/architecture/agent-state.md) and means a consumer that missed an
// observation is never left with a stale partial picture.
type TelemetryHealth struct {
	// Target is the egress target this describes: "http" or "eventhub".
	Target string `json:"target"`

	// QueuedBatches, QueuedEvents, and QueuedBytes describe the backlog
	// awaiting redelivery. All three are zero on a healthy target.
	QueuedBatches int   `json:"queuedBatches"`
	QueuedEvents  int   `json:"queuedEvents"`
	QueuedBytes   int64 `json:"queuedBytes"`

	// OldestAgeMs is how long the oldest undelivered batch has been waiting.
	// Age matters independently of size: a small backlog that has not moved
	// in an hour is a different failure from a large one draining steadily.
	OldestAgeMs int64 `json:"oldestAgeMs"`

	// MaxAttempts is the highest redelivery attempt count of any queued
	// batch. Rising attempts with a flat OldestAgeMs distinguish "keeps
	// failing" from "waiting out one long backoff".
	MaxAttempts int `json:"maxAttempts,omitempty"`

	// Stuck is true when OldestAgeMs has reached StuckAfterMs: the queue
	// has not fully drained within the configured window. It is reported on
	// the transition in each direction and is independent of size, so a
	// backlog too small to cross a size notch still escalates once it is
	// old enough.
	Stuck        bool  `json:"stuck,omitempty"`
	StuckAfterMs int64 `json:"stuckAfterMs,omitempty"`

	// QuarantinedEvents and QuarantinedBytes count, cumulatively for the
	// process, the events written to the quarantine file instead of being
	// sent: content the transport could never carry under the configured
	// oversize policy. They are on disk, not in the stream, and every
	// increase is reported.
	QuarantinedEvents int   `json:"quarantinedEvents,omitempty"`
	QuarantinedBytes  int64 `json:"quarantinedBytes,omitempty"`

	// SoftWarnBytes is the advisory threshold the crossing percentages are
	// measured against, echoed so a consumer can render "180 MB of 500 MB"
	// without knowing the engine's configuration.
	SoftWarnBytes int64 `json:"softWarnBytes"`

	// PercentOfSoftWarn is QueuedBytes as a percentage of SoftWarnBytes. It
	// is not clamped: a queue past its advisory threshold reports above 100,
	// which is meaningful rather than an error, since the queue is unbounded
	// by default and legitimately keeps growing.
	PercentOfSoftWarn int `json:"percentOfSoftWarn"`

	// CrossedThreshold is the escalation step this observation represents
	// (50, 75, 85, or 95), or 0 when the observation is not a crossing.
	// Consumers that only want to alert on escalation filter on this.
	CrossedThreshold int `json:"crossedThreshold,omitempty"`

	// Healthy is false when the last delivery attempt failed. A target can
	// be unhealthy with an empty queue (the failure just happened) or
	// healthy with a non-empty one (draining a past outage).
	Healthy bool `json:"healthy"`

	// LastError is the most recent delivery error, empty when healthy.
	LastError string `json:"lastError,omitempty"`

	// Critical marks a condition the engine cannot resolve by retrying —
	// today, a queue write that failed because the disk is full. Durability
	// is genuinely lost at that point: events are neither delivered nor
	// persisted, so this is the one health condition that is an error rather
	// than a backlog report.
	Critical bool `json:"critical,omitempty"`
}

// HealthObserver receives a TelemetryHealth snapshot. A host installs one to
// project queue state onto its own surfaces; nil (the default) means the
// engine keeps its log-only behavior.
type HealthObserver func(TelemetryHealth)

// healthTracker holds the per-queue state needed to decide when an
// observation is worth reporting. Reporting every enqueue would drown a
// consumer during exactly the outage it needs to reason about, so crossings
// escalate upward and reset only once the queue actually drains, stuck is
// reported on each transition, and quarantine on each increase.
type healthTracker struct {
	mu              sync.Mutex
	observer        HealthObserver
	lastNotch       int
	lastStuck       bool
	lastQuarantined int
	lastReport      time.Time
}

// notchFor returns the highest warning fraction the given percentage has
// reached, or 0 below the first one.
func notchFor(percentOfSoftWarn int) int {
	notch := 0
	for _, f := range warnFractions {
		if percentOfSoftWarn >= f {
			notch = f
		}
	}
	return notch
}

// observe reports a health snapshot when it is worth reporting: a new
// escalation notch, a drop back to healthy after a backlog, a stuck
// transition in either direction, newly quarantined events, or a critical
// condition. Steady-state growth between notches stays silent.
func (t *healthTracker) observe(h TelemetryHealth) {
	if t == nil {
		return
	}
	t.mu.Lock()
	observer := t.observer
	notch := notchFor(h.PercentOfSoftWarn)
	crossed := notch > t.lastNotch
	recovered := notch == 0 && t.lastNotch > 0
	stuckChanged := h.Stuck != t.lastStuck
	quarantined := h.QuarantinedEvents > t.lastQuarantined
	report := h.Critical || crossed || recovered || stuckChanged || quarantined
	if report {
		if crossed {
			h.CrossedThreshold = notch
		}
		t.lastNotch = notch
		t.lastStuck = h.Stuck
		t.lastQuarantined = h.QuarantinedEvents
		t.lastReport = time.Now()
	}
	t.mu.Unlock()

	if report && observer != nil {
		observer(h)
	}
}

// isDiskFull reports whether an error is a no-space-left condition, the one
// queue-write failure the engine treats as critical rather than transient.
// Checked structurally rather than by string match so it holds across
// platforms and wrapped errors.
func isDiskFull(err error) bool {
	if err == nil {
		return false
	}
	var errno syscall.Errno
	if errors.As(err, &errno) {
		return errno == syscall.ENOSPC || errno == syscall.EDQUOT
	}
	var pathErr *fs.PathError
	if errors.As(err, &pathErr) {
		return isDiskFull(pathErr.Err)
	}
	return false
}

// SetHealthObserver installs fn as the receiver for every retry queue this
// Collector owns. A host calls this once after construction to project
// delivery health onto its own surfaces (a typed engine event, a client
// notification). Passing nil restores the default log-only behavior.
//
// Safe on a Collector with no network target: the queues are nil and the
// call is a no-op, so a caller need not first ask which targets exist.
func (c *Collector) SetHealthObserver(fn HealthObserver) {
	if c == nil {
		return
	}
	c.httpRetry.setHealthObserver(fn)
	c.eventHubRetry.setHealthObserver(fn)
}

// TelemetryHealthSnapshot reports the current delivery health of every
// configured target, for a caller that wants state on demand rather than on
// transition — a status command, or a client reconnecting mid-outage that
// needs the current picture rather than waiting for the next crossing.
func (c *Collector) TelemetryHealthSnapshot() []TelemetryHealth {
	if c == nil {
		return nil
	}
	var out []TelemetryHealth
	if c.httpRetry != nil {
		out = append(out, c.httpRetry.stats(true, ""))
	}
	if c.eventHubRetry != nil {
		out = append(out, c.eventHubRetry.stats(true, ""))
	}
	return out
}
